### PR TITLE
Add highest node count

### LINKED ISSUE #44: Add highest seen node count to Telemetry UI
Currently, seeing the live node count is great for the telemetry experience. However, for marketing and community, this makes it difficult to gauge where was the last peek node count was as nodes can come and go offline.

We want to add a element that shows what was the last highest seen node count and track that.

### PR DESCRIPTION
**Pull request by isSerge**
image

Closes #44

### CODE REVIEW

**Review by nazar-pc — COMMENTED** [review]
Makes sense overall, but it'll only last until restart, which I guess isn't quite what we want?

**Comment by isSerge** [comment]
> Makes sense overall, but it'll only last until restart, which I guess isn't quite what we want?

We need this feature today (as we discussed on the App sync yesterday), so I believe this is the best we can do now

**Comment by i1i1** [comment]
> Makes sense overall, but it'll only last until restart, which I guess isn't quite what we want?

`telemetry_core` is totally stateless, I'm not sure about adding local database just for this feature. WDUT?

**Comment by nazar-pc** [comment]
I think database might be an overkill, but saving at least a JSON somewhere when max number gets updated would be helpful I think.

File: backend/telemetry_core/src/aggregator/aggregator_set.rs [hunk 1]
```diff
@@ -19,12 +20,18 @@ impl AggregatorSet {
...
         opts: AggregatorOpts,
     ) -> anyhow::Result<AggregatorSet> {
         assert_ne!(num_aggregators, 0, "You must have 1 or more aggregator");
 
         let aggregators = futures::future::try_join_all(
-            (0..num_aggregators).map(|_| Aggregator::spawn(opts.clone())),
+            std::iter::once(AggregatorOpts {
+                metadata_path,
+                ..opts.clone()
+            })
```
Comment (nazar-pc) [inline backend/telemetry_core/src/aggregator/aggregator_set.rs:32]: Hm... does this mean multiple aggregators will concurrently write to disk? We don't want that, we should only have one that does writes.

File: backend/telemetry_core/src/state/batched.rs [hunk 2]
```diff
@@ -88,6 +140,16 @@ impl State {
...
     pub fn drain_updates_for_all_feeds(&mut self) -> FeedMessageSerializer {
+        self.metadata.update(&self.chains);
+        if let Some(path) = &self.metadata_path {
+            if let Err(err) = serde_json::to_vec(&self.metadata)
+                .map_err(anyhow::Error::from)
+                .and_then(|bytes| std::fs::write(path, bytes).map_err(anyhow::Error::from))
+            {
+                log::error!("Failed to save metadata: {err}");
+            }
+        }
```
Comment (nazar-pc) [inline backend/telemetry_core/src/state/batched.rs]: Ideally we'd check if it has changed before writing. If it didn't, no need to write an update.

File: docker-compose.yml [hunk 3]
```diff
@@ -25,7 +25,7 @@ services:
       - /var/run:uid=101,gid=101
       - /app/tmp:uid=101,gid=101
     environment:
-      SUBSTRATE_TELEMETRY_URL: wss://telemetry.subspace.network/feed
+      SUBSTRATE_TELEMETRY_URL: wss://telemetry.subspace.network/feed
```
Comment (nazar-pc) [inline docker-compose.yml]: Accidental whitespace

File: docker-compose.yml [hunk 4]
```diff
@@ -67,8 +67,18 @@ services:
...
       - 127.0.0.1:8000:8000
     expose:
       - 8000
+
+volumes:
+  # Just create it with:
+  # $ docker volume create telemetry-core-metadata
+  # $ docker run -it -v telemetry-core-metadata:/m alpine chown 1000:1000 /m 
+  telemetry-core-metadata:
+    external: true
```
Comment (nazar-pc) [inline docker-compose.yml]: I'd say remove `external: true` and let it be managed by Docker Compose, much easier that way.
Then you can set permissions in Dockerfile and avoid manual actions, here is an example: https://github.com/subspace/subspace/blob/7b1412208889c3bf52aace9ea3d56dc328645eba/Dockerfile-node#L65-L67

**Review by nazar-pc — COMMENTED** [review]
Makes sense otherwise

File: backend/telemetry_core/src/aggregator/aggregator_set.rs (hunk 1 above)
Comment (i1i1) [reply backend/telemetry_core/src/aggregator/aggregator_set.rs:32]: Nope, only the first one. This is why the first one has `metdata_path` set and the rest of aggregators have it as `None`.

Comment (nazar-pc) [reply backend/telemetry_core/src/aggregator/aggregator_set.rs:32]: I see now, thanks. The old code with a simple iterator was much more readable for me, you could have had a simple `if` statement in there and leave the rest as is.

File: backend/Dockerfile [hunk 5]
```diff
@@ -24,7 +24,9 @@ RUN useradd -m -u 1000 -U telemetry && \
     apt-get -y install openssl && \
     apt-get autoremove -y && \
     apt-get clean && \
-    rm -rf /var/lib/apt/lists/
+    rm -rf /var/lib/apt/lists/ && \
+    mkdir /var/telemetry_core && \
```
Comment (nazar-pc) [inline backend/Dockerfile:28]: I believe this should also be exposed as a volume like in the link I provided earlier, may or may not work properly without it.

File: backend/telemetry_core/src/state/batched.rs [hunk 6]
```diff
@@ -41,10 +41,18 @@ struct Metadata {
...
+        &mut self,
+        chains: impl IntoIterator<Item = (&'a BlockHash, &'a ChainUpdates)>,
+    ) -> bool {
+        let mut updated = false;
         for (hash, chain) in chains {
-            self.chains.entry(*hash).or_default().highest_node_count = chain.highest_node_count;
+            let entry = self.chains.entry(*hash);
+            updated |= matches!(entry, Entry::Vacant(_))
+                | matches!(&entry, Entry::Occupied(entry) if entry.get().highest_node_count != chain.highest_node_count);
+            entry.or_default().highest_node_count = chain.highest_node_count;
```
Comment (nazar-pc) [inline backend/telemetry_core/src/state/batched.rs]: This is very not readable code to me!
Just use match on `entry` instead.
And PLEASE, avoid bitwise or.

I believe you wanted to write this:
```rust
match self.chains.entry(*hash) {
    Entry::Vacant(entry) => {
        entry.insert(*chain);
        updated = true;
    }
    Entry::Occupied(entry) => {
        if entry.get() != chain {
            entry.insert(*chain);
            updated = true;
        }
    }
}
```

Yes, it is more code, but I can personally read it without thinking, while with your version I need to think quite hard. Also I'd not use `impl IntoIterator` unless you have strong reason to, you just make code more complex than it needs to be with extra lifetime argument and such.

TL;DR: Don't try to write smart code, write simple code that is easy to review.

File: backend/Dockerfile (hunk 5 above)
Comment (i1i1) [reply backend/Dockerfile:28]: For a record, it actually works even without it.

**Review by nazar-pc — DISMISSED** [review]

**Review by i1i1 — APPROVED** [review]

**Review by nazar-pc — APPROVED** [review]
